rotkehlchen/chain/evm/decoding/balancer/decoder.py

```diff
@@ -0,0 +1,220 @@
+            pool_cache_type: CacheType,
+            read_fn: Callable[[ChainID], tuple[set[ChecksumEvmAddress], set[ChecksumEvmAddress]]],
+            counterparty: Literal['balancer-v1', 'balancer-v2'],
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        ReloadablePoolsAndGaugesDecoderMixin.__init__(
+            self,
+            evm_inquirer=evm_inquirer,
+            cache_type_to_check_for_freshness=pool_cache_type,
+            query_data_method=query_balancer_data,
+            read_data_from_cache_method=read_fn,
+            chain_id=evm_inquirer.chain_id,
+        )
+        self.counterparty: Literal['balancer-v1', 'balancer-v2'] = counterparty
+
+    @property
+    def pools(self) -> set[ChecksumEvmAddress]:
+        assert isinstance(self.cache_data[0], set), f'{self.counterparty} Decoder cache_data[0] is not a set'  # noqa: E501
+        return self.cache_data[0]
+
+    def _decode_gauge_events(self, context: DecoderContext) -> DecodingOutput:
+        if context.tx_log.topics[0] not in (GAUGE_DEPOSIT_TOPIC, GAUGE_WITHDRAW_TOPIC):
+            return DEFAULT_DECODING_OUTPUT
+
+        gauge_asset = self.base.get_or_create_evm_token(context.tx_log.address)
+        amount = asset_normalized_value(
+            amount=int.from_bytes(context.tx_log.data[:32]),
+            asset=gauge_asset,
+        )
+        from_event_type, from_event_subtype, to_event_type, to_event_subtype, to_notes = None, None, None, None, ''  # noqa: E501
+        for event in context.decoded_events:
+            if (
+                event.balance.amount == amount and
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                (evm_asset := event.asset.resolve_to_evm_token()).evm_address in self.pools
+            ):
+                from_event_type = HistoryEventType.RECEIVE
+                from_event_subtype = HistoryEventSubType.NONE
+                to_event_type = HistoryEventType.RECEIVE
+                to_event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                to_notes = f'Receive {amount} {gauge_asset.symbol} after depositing in balancer gauge'  # noqa: E501
+
+                event.counterparty = self.counterparty
+                event.event_type = HistoryEventType.DEPOSIT
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.sequence_index = self.base.get_sequence_index(context.tx_log)
+                event.notes = f'Deposit {amount} {evm_asset.symbol} into balancer gauge'
+            elif (
+                event.balance.amount == amount and
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                (evm_asset := event.asset.resolve_to_evm_token()).evm_address in self.pools
```

Comment:
it's an if/elif block so the evm_asset variable is scoped to the block and only one of the conditionals get executed for each loop.

or am i misunderstanding?

Reply:
since you have the event types as filters it shouldn't happen that you request it twice sicne if it falls in one of the branches it won't reach that point in the other but... why not avoid repeating things? You can add before this two branches

``` 
if (
     event.balance.amount != amount or
     event.event_subtype != HistoryEventSubType.NONE or
     (evm_asset := event.asset.resolve_to_evm_token()).evm_address not in self.pools
):
    continue
```